sdk/state/update_test.go

```diff
@@ -0,0 +1,75 @@
+package state
+
+import (
+	"testing"
+
+	"github.com/stellar/go/keypair"
+	"github.com/stellar/go/network"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestLastConfirmedPayment(t *testing.T) {
+	localSigner := keypair.MustRandom()
+	remoteSigner := keypair.MustRandom()
+	localEscrowAccount := &EscrowAccount{
+		Address:        keypair.MustRandom().FromAddress(),
+		SequenceNumber: int64(101),
+	}
+	remoteEscrowAccount := &EscrowAccount{
+		Address:        keypair.MustRandom().FromAddress(),
+		SequenceNumber: int64(202),
+	}
+	sendingChannel := NewChannel(Config{
+		NetworkPassphrase:   network.TestNetworkPassphrase,
+		Initiator:           true,
+		LocalSigner:         localSigner,
+		RemoteSigner:        remoteSigner.FromAddress(),
+		LocalEscrowAccount:  localEscrowAccount,
+		RemoteEscrowAccount: remoteEscrowAccount,
+	})
+	receiverChannel := NewChannel(Config{
+		NetworkPassphrase:   network.TestNetworkPassphrase,
+		Initiator:           false,
+		LocalSigner:         remoteSigner,
+		RemoteSigner:        localSigner.FromAddress(),
+		LocalEscrowAccount:  remoteEscrowAccount,
+		RemoteEscrowAccount: localEscrowAccount,
+	})
+
+	p, err := sendingChannel.ProposePayment(Amount{
+		Asset:  NativeAsset{},
+		Amount: 200,
+	})
+	require.NoError(t, err)
+
+	p, fullySigned, err := receiverChannel.ConfirmPayment(p)
+	assert.False(t, fullySigned)
+	require.NoError(t, err)
+	assert.Equal(t, p, receiverChannel.latestUnconfirmedPayment)
+
+	// Confirming a payment with same sequence number but different Amount should error
+	pDifferent := Payment{
+		IterationNumber: 1,
+		Amount: Amount{
+			Asset:  NativeAsset{},
+			Amount: 400,
+		},
+		CloseSignatures: p.CloseSignatures,
+	}
+	_, fullySigned, err = receiverChannel.ConfirmPayment(pDifferent)
+	assert.False(t, fullySigned)
+	require.Error(t, err)
+	require.Equal(t, "a different unconfirmed payment exists", err.Error())
+
+	// Confirming a payment with same sequence number and same amount should pass
+	p, fullySigned, err = sendingChannel.ConfirmPayment(p)
+	assert.True(t, fullySigned)
+	require.NoError(t, err)
+	assert.Equal(t, Payment{}, sendingChannel.latestUnconfirmedPayment)
+
+	p, fullySigned, err = receiverChannel.ConfirmPayment(p)
```

Comment:
What are you testing here? That confirming payments is an idempotent operation? Just wonder if it's intentional that we're not triggering an error here because the tx has already been confirmed.

Reply:
ya  this isn't obvious but the receiver actually has to call `ConfirmPayment` twice, so this is testing that. The first time called it adds its signatures. The second time called it essentially is just grabbing the signatures the sender gave. So essentially:

1. Sender calls ProposePayment
2. Receiver calls ConfirmPayment
3. Sender calls ConfirmPayment
4. Receiver calls ConfirmPayment

This was a design decision to use the same method for steps 2,3,4 since they're similar

Reply:
I added some comments that explains that:
https://github.com/stellar/experimental-payment-channels/pull/81/commits/df2993648747a22ee09d8f2ba9f356e3c431b0b4